DRILL-6461: Added basic data correctness tests for hash agg, and improved operator unit testing framework.

ilooner opened the pull request:
This change:

 1. HashAgg previously had no data correctness tests, so this test adds some in order to have better test coverage for the HashAgg refactoring changes.
 2. Improves the OperatorTestBuilder framework to work with rowsets.

ilooner commented:
@sachouche @sohami please review.

sachouche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java:
```diff
@@ -1175,6 +1175,7 @@ public AggIterOutcome outputCurrentBatch() {
       v.getValueVector().getMutator().setValueCount(numOutputRecords);
     }
 
+    outContainer.setRecordCount(numOutputRecords);
```
Noticed the old code is using INFO debug level; shouldn't it be using debug?

ilooner replied:
- I created an issue for centralizing setting the record counts here. We can probably just make VectorContainer.setRecordCount set the record count for value vectors as well instead of just itself. https://issues.apache.org/jira/browse/DRILL-6675
 - Discussed offline, I'll make the logging changes to HashAgg as part of https://issues.apache.org/jira/browse/DRILL-6253

sachouche replied:
Thank you Tim for the due diligence!

sachouche commented on exec/java-exec/src/main/java/org/apache/drill/exec/record/RecordBatchSizer.java:
```diff
@@ -639,6 +641,8 @@ public ColumnSize getColumn(String name) {
   // columns can be obtained from children of topColumns.
   private Map<String, ColumnSize> columnSizes = CaseInsensitiveMap.newHashMap();
 
+  private List<ColumnSize> columnSizesList = new ArrayList<>();
```
Why do this? the columnSizes map could return a Collection<ColumnSize> (through java.util.Map.values()) efficiently  (that is, in place).

ilooner replied:
It's for convenience. VectorContainers expose their ValueVectors based on index, so it is handy to be able to lookup column size based on the same index as well. Otherwise I would have to get the Materialized field for the a column, extract the name, and then lookup the ColumnSize in the map.

sachouche replied:
I see; can you please add a comment indicating this behavior (especially within the columnsList() API).

Thanks!

ilooner replied:
done.

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/record/VectorContainer.java:
```diff
@@ -42,6 +42,8 @@
   protected final List<VectorWrapper<?>> wrappers = Lists.newArrayList();
   private BatchSchema schema;
+  private SelectionVector2 selectionVector2;
+  private SelectionVector4 selectionVector4;
```
I don't think we should include SV2/SV4 inside a VectorContainer. They are more of RecordBatch level members. VectorContainer should only have metadata and set of vectors in it which is what it is today.

ilooner replied:
This has actually been a point of confusion for me. VectorContainer and RecordBatch both implement VectorAccessible, which has the following methods:

```
  SelectionVector2 getSelectionVector2();
  SelectionVector4 getSelectionVector4();
```

If the intention is that VectorContainers should never hold selection vectors, then the interfaces VectorContainer implements should reflect that. What are you're thoughts about refactoring the VectorAccessable to remove the selection vector methods? In the meantime I will remove the these changes from VectorContainer.

ilooner replied:
Cool made https://issues.apache.org/jira/browse/DRILL-6683 to track that change.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/rowSet/RowSetBatch.java:
```diff
@@ -29,13 +30,28 @@
+/**
+ * A mock operator that returns the provided {@link RowSet}s as batches. Currently it's assumed that all the {@link RowSet}s have the same schema.
+ */
+public class RowSetBatch implements CloseableRecordBatch {
```
RowSetBatch predates MockRecordBatch :/ . But I'll see if I can make MockRecordBatch play nice with RowSets. Since MockRecordBatch is used for many many tests, I will make it work with RowSets while keeping some deprecated constructors for VectorContainers so that I don't have to update every test.

sohami replied:
If you want you can leave it as is for now. I can take a look on how to refactor MockRecordBatch class along with other tests which uses it, which I have to do anyways. I was also thinking of making it use RowSets instead of container. 
The reason for creating separate class was to use it specifically for testing different IterOutcomes in context of Lateral&Unnest only whereas RowSetBatch looked more of a very simple implementation to just hold one container. Was not sure if MockRecordBatch will play well with other operators in general.

ilooner replied:
@sohami It's okay, I've already made the move to MockRecordBatch for this PR. Just doing a final test runs, and debugging straggling issues.

ilooner commented:
@sohami I removed the **RowSetBatch** and made the **MockRecordBatch** work with RowSets. Please take another look. 

This proved to be a fruitful exercise because using MockRecordBatch exposed some bugs and corner cases in the vectors, and then those bugs and corner cases lead to even more bugs being exposed. The specific issues were:

 - Previously I assumed **copyEntry** was a safe operation (does reallocs when needed), but this was not entirely true. **copyEntry** was safe for non-empty VarLength vectors, and Nullable vectors copying from another Nullable vector. It was not safe for empty VarLength vectors, copying from a non-nullable vector to a nullable vector, or fixed length vectors. So Included fixes for these three corner cases.
 - I made **copyEntry** safe for empty VarLength vectors by always making sure the first entry in the offset vector existed and was zero on initialization of the vector. This caused memory to be leaked because some operators created empty VectorContainers and never closed them:
  - TopN created a dummy VectorContainer for no reason and never closed it, so I removed it.
  - HashTableTemplate created a dummy template VectorContainer and never closed it, so I close it now. We should actually probably avoid using the dummy VectorContainer entirely and just use a schema, but I would rather do that in a separate change.

ilooner commented:
Arg! There appears to be one random test failing on Travis. I will look into it.

ilooner commented:
@sohami The PR is ready for another round of review.

I removed the fix that allowed **copyEntry** to work with empty VarLength vectors since value vectors should not be manipulated unless allocateNew is called. I fixed the failing test by calling allocateNew on all the vectors in the destination vector container.

sohami commented on exec/java-exec/src/test/java/org/apache/drill/test/OperatorTestBuilder.java:
```diff
@@ -0,0 +1,300 @@
+      if (combineOutputBatches) {
+        final RowSet expectedBatch = expectedResults.get(0);
+        final RowSet actualBatch = DirectRowSet.fromSchema(
+          physicalOpUnitTestBase.operatorFixture.allocator, expectedBatch.schema());
```
shouldn't this be `actualResults.get(0).container().schema()` instead of `expectedBatch.schema()` ?

ilooner replied:
Changed.

sohami commented on exec/java-exec/src/test/java/org/apache/drill/test/OperatorTestBuilder.java:
```diff
@@ -0,0 +1,300 @@
+        return;
+      }
+
+      if (combineOutputBatches) {
```
what is the idea behind `combineOutputBatches` since looks like we are just copying data from actual received vector containers to a single container ?

ilooner replied:
This is needed for operators like HashAgg. Internally HashAgg maintains multiple partitions, and on completion separate output batches are emitted for each partition. So if HashAgg is using 32 partitions internally, it will emit at least 32 batches even for a small test data set.

It is difficult to predict what records go to what partitions without going through the entire hashcode calculation for the test. Additionally I only care about the correctness of the final results, and not about what partition got what records. So combining all the results into a single batch for comparison simplifies testing.

sohami replied:
Thanks for explanation. But in case of combining batches I am not seeing actual comparison happening between actual and expected batch.

ilooner replied:
Thanks for catching this. It looks like I deleted the compare statement during a refactor. This is a really stupid mistake, I'll add a test for OperatorTestBuilder to catch it.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/MockRecordBatch.java:
```diff
@@ -158,19 +210,34 @@ public IterOutcome next() {
+          final SelectionVector2 inputSv2 = ((RowSet.SingleRowSet) rowSet).getSv2();
+
+          if (sv2 != null) {
+            // Operators like PartitionLimitRecordBatch assume that new values for an Sv2 are transferred in.
```
done

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/OperatorTestBuilder.java:
```diff
@@ -0,0 +1,300 @@
+  private void validate() {
+    if (combineOutputBatches) {
+      Preconditions.checkArgument(expectedResults.isEmpty() || expectedResults.size() == 1,
+        "The number of expected result batches needs to be zero or one when combining input batches");
```
fixed thanks for catching

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/OperatorTestBuilder.java:
```diff
@@ -0,0 +1,300 @@
+    if (combineOutputBatches) {
+      Preconditions.checkArgument(expectedResults.isEmpty() || expectedResults.size() == 1,
+        "The number of expected result batches needs to be zero or one when combining input batches");
+      Preconditions.checkArgument((expectedResults.isEmpty() && (!expectedNumBatchesOpt.isPresent() && expectedTotalRowsOpt.isPresent())) ||
```
Fixed defined to define.

The preconditions are different. Specifically these two parts of the conditions.

```
(!expectedNumBatchesOpt.isPresent() && expectedTotalRowsOpt.isPresent())
(expectedNumBatchesOpt.isPresent() || expectedTotalRowsOpt.isPresent())
```

They are different because when you are combining output batches it doesn't make sense to check the number of output batches in the test.

sohami replied:
makes sense. Thanks for explanation.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/OperatorTestBuilder.java:
```diff
@@ -0,0 +1,300 @@
+
+      int actualTotalRows = actualResults.stream()
+        .map(rowSet -> rowSet.rowCount())
+        .reduce((Integer countA, Integer countB) -> countA + countB)
```
done

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/OperatorTestBuilder.java:
```diff
@@ -0,0 +1,300 @@
+            rowSetContainer.setRecordCount(actualRowSet.rowCount());
+
+            switch (actualRowSet.indirectionType()) {
+              case NONE:
```
Why? The code doesn't support output batches with a selection vector yet. I'd rather add support for it if it's actually needed.

sohami replied:
But you are adding corresponding RowSet inside actualResults [here](https://github.com/apache/drill/pull/1344/files/332b5f34d43f8f546824619c966b978eb5788b47#diff-de4ffc656d08e47bd1b8c716dca26f3eR109) based on output container has SV or not.

ilooner replied:
Yeah I guess I did (forgot it was so long ago). I don't remember the details, but I added the code you linked to because it was easy to add. When implementing the copy here I ran into an issue (I don't quite remember what, it's been a couple months). Since handling sv2s wasn't required for testing HashAgg I gave up on it decided to leave it unimplemented. 

I'd rather not pursue completing the code to handle sv2s in this PR. The intention here was to test HashAgg, if somebody needs to handle output batches with selection vectors for another operator, they can add it as a separate PR. I'll remove the code snippet you've referenced and throw an UnsupportedException there as well. I made a jira to track this feature https://issues.apache.org/jira/browse/DRILL-6698 and will reference the jira number in the code.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/svremover/AbstractGenericCopierTest.java:
```diff
@@ -20,77 +20,106 @@
+      final VectorContainer destContainer = new VectorContainer(allocator, batchSchema);
+
+      for (VectorWrapper vectorWrapper: destContainer) {
+        vectorWrapper.getValueVector().allocateNew();
```
Thanks for catching this. I've moved the allocation code into a static method and call it in GenericCopier.copyRecords.

It is still necessary to allocate the vectors in the appendRecords test though, since appendRecords assumes the vector container is pre existing.

ilooner commented:
@sohami Applied review comments. Please take another look.

ilooner commented:
@sohami Addressed review comments. Please take another look.

I also implemented some tests for OperatorTestBuilder. The simple tests used the project operator to pass through input batches. While implementing those tests I discovered a bug in Project where the recordCount is not set on output containers, so I fixed that as part of this PR.
